### PR TITLE
Added SFC Service, all calls.

### PR DESCRIPTION
**Pull request by dmitryintel**
Adding SFC functionality.

### CODE REVIEW

File: core/src/main/java/org/openstack4j/api/networking/NetworkingService.java [hunk 1]
```diff
@@ -78,9 +79,14 @@
...
+
     /**
      * @return the Availability Zone Service API
      */
     AvailabilityZoneService availabilityzone();
+
+    /**
+     * @return the Service Function Chain Service API
+     */
+    SFCService sfc();
```
Comment (arvindn05) [inline core/src/main/java/org/openstack4j/api/networking/NetworkingService.java]: i would use the expanded form ServiceFunctionChainService for the class name atleast.

File: core/src/main/java/org/openstack4j/openstack/networking/domain/ext/NeutronFlowClassifier.java [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+
+import org.openstack4j.model.network.ext.FlowClassifier;
+import org.openstack4j.model.network.ext.builder.FlowClassifierBuilder;
+import org.openstack4j.openstack.common.ListResult;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.annotation.JsonRootName;
+
+/**
+ *
```
Comment (arvindn05) [inline core/src/main/java/org/openstack4j/openstack/networking/domain/ext/NeutronFlowClassifier.java]: nitpick but too many blank/comment lines?

File: core/src/main/java/org/openstack4j/openstack/networking/domain/ext/NeutronFlowClassifier.java [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+	private static final long serialVersionUID = 1L;
+
+	@JsonProperty
+	private String id;
+
+	@JsonProperty
+	private String name;
+
+	@JsonProperty("project_id")
+	private String tenantId;
```
Comment (arvindn05) [inline core/src/main/java/org/openstack4j/openstack/networking/domain/ext/NeutronFlowClassifier.java]: use project id for the field name as well. Openstack replaced all references of tenant -> project

File: core/src/main/java/org/openstack4j/openstack/networking/domain/ext/NeutronFlowClassifier.java [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+        private List<NeutronFlowClassifier> flowClassifiers;
+
+        public List<NeutronFlowClassifier> value() {
+            return flowClassifiers;
+        }
+    }
+
+	public static class FlowClassifierConcreteBuilder implements FlowClassifierBuilder {
+
+		private NeutronFlowClassifier m;
```
Comment (arvindn05) [inline core/src/main/java/org/openstack4j/openstack/networking/domain/ext/NeutronFlowClassifier.java]: use more appropriate name? like neutronFlowClassifier

File: core/src/main/java/org/openstack4j/openstack/networking/internal/ext/SFCServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+import java.util.List;
+
+import org.openstack4j.api.networking.ext.SFCService;
+import org.openstack4j.model.network.ext.FlowClassifier;
+import org.openstack4j.openstack.networking.domain.ext.NeutronFlowClassifier.FlowClassifiers;
+import org.openstack4j.openstack.networking.internal.BaseNetworkingServices;
+
+/**
+ * Tests Compute -> Service Function Chain API against the mock webserver and spec based
```
Comment (arvindn05) [inline core/src/main/java/org/openstack4j/openstack/networking/internal/ext/SFCServiceImpl.java]: tests?

**Comment by arvindn05** [comment]
@dmitryintel looks like lots of whitespace changes? revert the whitespace changes?

**Comment by auhlig** [comment]
Thanks a lot for the PR @dmitryintel and also thanks for reviewing @arvindn05

File: core/src/main/java/org/openstack4j/api/Apis.java [hunk 6]
```diff
@@ -1,224 +1,234 @@
...
-
-
-    private static APIProvider initializeProvider() {
-        // No need to check for emptiness as there is default implementation registered
-        APIProvider p = ServiceLoader.load(APIProvider.class, Apis.class.getClassLoader()).iterator().next();
-        p.initialize();
-        return p;
-    }
-}
+package org.openstack4j.api;
```
Comment (auhlig) [inline core/src/main/java/org/openstack4j/api/Apis.java]: Why does the whole file change?

File: core/src/main/java/org/openstack4j/api/OSClient.java [hunk 7]
```diff
@@ -1,350 +1,357 @@
...
-
-    /**
-     * Returns the DNS Service API
-     *
-     * @return the DNS service
-     */
-    DNSService dns();
-
-}
+package org.openstack4j.api;
```
Comment (auhlig) [inline core/src/main/java/org/openstack4j/api/OSClient.java]: Why does the whole file change?

File: core/src/main/java/org/openstack4j/openstack/networking/domain/ext/NeutronFlowClassifier.java (hunk 3 above)
Comment (auhlig) [reply core/src/main/java/org/openstack4j/openstack/networking/domain/ext/NeutronFlowClassifier.java]: Agreed. tenant only causes confusion

File: core/src/main/java/org/openstack4j/openstack/networking/internal/NetworkingServiceImpl.java [hunk 8]
```diff
@@ -18,7 +18,7 @@
 
 /**
  * OpenStack Networking Operations API
- * 
+ *
```
Comment (auhlig) [inline core/src/main/java/org/openstack4j/openstack/networking/internal/NetworkingServiceImpl.java]: revert tab/whitespace changes in this file

**Comment by dmitryintel** [comment]
TenantId comes from Resource interface. I can change that, but that would entail a change on multiple classes. I renamed the field without renaming the getter.

File: core/src/main/java/org/openstack4j/openstack/networking/internal/NetworkingServiceImpl.java (hunk 8 above)
Comment (dmitryintel) [reply core/src/main/java/org/openstack4j/openstack/networking/internal/NetworkingServiceImpl.java]: TenantId comes from Resource interface. I can change that, but that would entail a change on multiple classes outside of the scope for this enhancement. I renamed the field without renaming the getter.

File: core/src/main/java/org/openstack4j/model/network/ext/builder/FlowClassifierBuilder.java [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+	 * @param name : Human readable name for the flow classifier
+	 * @return FlowClassifierBuilder
+	 */
+	FlowClassifierBuilder name(String name);
+
+	/**
+	 * @param tenantId : Tenant (project) identifier
+	 * @return FlowClassifierBuilder
+	 */
+	FlowClassifierBuilder tenandId(String tenantId);
```
Comment (arvindn05) [inline core/src/main/java/org/openstack4j/model/network/ext/builder/FlowClassifierBuilder.java]: use project id?

File: core-test/src/main/resources/network/sfc/flow_classifiers.json [hunk 10]
```diff
@@ -0,0 +1,36 @@
+{
+    "flow_classifiers": [
+        {
+            "name": "FC1",
+            "project_id": "12345678909876543210abcdefabcdef",
+            "tenant_id": "12345678909876543210abcdefabcdef",
```
[the remark this answers is not in the document]
Comment (dmitryintel) [reply core-test/src/main/resources/network/sfc/flow_classifiers.json]: They are both returned from the actual openstack

File: core-test/src/main/resources/network/sfc/flow_classifier.json [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+        "description": "Openstack4j test flow classifier 1",
+        "protocol": "TCP",
+        "ethertype": "IPv4",
+        "source_port_range_min": 22,
+        "source_port_range_max": 3000,
+        "destination_port_range_min": 80,
+        "destination_port_range_max": 80,
+        "source_ip_prefix": null,
+        "destination_ip_prefix": "10.20.30.40",
+        "logical_source_port": "abcdefab-abcd-abcd-abcd-abcdefabcdef",
```
Comment (arvindn05) [inline core-test/src/main/resources/network/sfc/flow_classifier.json]: logical destination port is missing

**Comment by dmitryintel** [comment]
@arvindn05 @auhlig  all my intended changes are in

File: core-test/src/main/resources/network/sfc/port_chain.json [hunk 12]
```diff
@@ -0,0 +1,12 @@
+{
+    "port_chain": {
+        "id" : "0abcdef",
+        "name": "PC1",
+        "project_id": "12345678909876543210abcdefabcdef",
+        "tenant_id": "12345678909876543210abcdefabcdef",
+        "description": "Openstack4j test port chain",
+        "chain_id": "43210",
+        "flow_classifiers":  [ "12345678-abcd-abcd-abcd-abcdefabcdef" ],
```
Comment (arvindn05) [inline core-test/src/main/resources/network/sfc/port_chain.json]: chain_parameters missing

File: core-test/src/main/resources/network/sfc/port_chain.json [hunk 13]
```diff
@@ -0,0 +1,12 @@
...
+        "port_chain_groups": [ "abcdefab-abcd-abcd-abcd-abcdefabcdef"]
```
Comment (arvindn05) [inline core-test/src/main/resources/network/sfc/port_chain.json]: should be port_pair_groups right? Is this payload from openstack or are we creating them by hand?

https://docs.openstack.org/networking-sfc/latest/contributor/api.html

File: core-test/src/main/resources/network/sfc/port_pair.json [hunk 14]
```diff
@@ -0,0 +1,11 @@
+{
+    "port_pair": {
+        "id" : "abcdef",
+        "name": "PP1",
+        "project_id": "12345678909876543210abcdefabcdef",
+        "tenant_id": "12345678909876543210abcdefabcdef",
+        "description": "Openstack4j test flow classifier 1",
+        "egress_id": "abcdefab-abcd-abcd-abcd-abcdefabcdef",
+        "ingress_id": "12345678-abcd-abcd-abcd-abcdefabcdef"
```
Comment (arvindn05) [inline core-test/src/main/resources/network/sfc/port_pair.json]: service_function_parameters missing

File: core-test/src/main/resources/network/sfc/port_pair_group.json [hunk 15]
```diff
@@ -1,13 +1,14 @@
 {
     "port_pair_group": {
         "id" : "abcdef",
+        "group_id": 42,
```
Comment (arvindn05) [inline core-test/src/main/resources/network/sfc/port_pair_group.json]: there is no group id for port pair groups

https://docs.openstack.org/networking-sfc/latest/contributor/api.html

Comment (dmitryintel) [reply core-test/src/main/resources/network/sfc/port_pair_group.json]: our openstack deployment returned it. I thought it was a gap in the docs.

**Comment by dmitryintel** [comment]
@auhlig , @arvindn05 would you be able to merge?

**Comment by arvindn05** [comment]
@dmitryintel we have to wait for the project maintainers to merge...i dont have merge permissions....hopefully @auhlig will be able to review and merge.

**Comment by vinodborole** [comment]
LGTM thanks for the review @arvindn05 and thanks for the contribution @dmitryintel

**Pull request merged**